src/borda_clj/client.clj

```diff
@@ -16,30 +17,57 @@
+    [max-buffer-size interval send on-send-error]
+    (let [measurements  (atom {})
+          running       (atom true)
+          submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
+          resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
+          flush         (fn [on-flush-error]
+                          (let [[m] (reset-vals! measurements {})]
+                            (if (not-empty m)
+                              (try
+                                (send m)
+                                (catch Throwable e
+                                  (on-flush-error m e))))))
+          stop          (fn [] (reset! running false) (flush on-send-error))]
+                                        ; periodically send to borda
+      (future (while @running (do
+                                (Thread/sleep interval)
+                                (flush (fn [m e] (on-send-error m e) (resubmit m))))))
+      [submit stop])))
+
+(defn reducing-submitter
   [max-buffer-size interval send on-send-error]
-  (let [measurements  (atom {})
-        running       (atom true)
-        submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
-        resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
-        flush         (fn [on-flush-error]
-                        (let [[m] (reset-vals! measurements {})]
-                          (if (not-empty m)
-                            (try
-                              (send m)
-                              (catch Throwable e
-                                (on-flush-error m e))))))
-        stop          (fn [] (reset! running false) (flush on-send-error))]
-        ; periodically send to borda
-    (future (while @running (do
-                              (Thread/sleep interval)
-                              (flush (fn [m e] (on-send-error m e) (resubmit m))))))
-    [submit stop]))
+  (let [report-ch (chan)]
+    (go-loop [measurements {}
+              flush-timeout (timeout interval)]
+      (let [flush (fn [on-error]
+                    (when (not-empty measurements)
+                      (try
+                        (send measurements)
+                        (catch Throwable e
+                          (on-error measurements e)))))
+            [[dimensions values :as v] ch] (alts! [report-ch flush-timeout])]
+          (cond (= ch flush-timeout)
+                (do
+                  (flush (fn [m e]
+                           (on-send-error m e)
+                           (onto-chan report-ch m false)))
+                  (recur {} (timeout interval)))
+                (some? v)
+                (recur (collect measurements max-buffer-size dimensions values)
+                       flush-timeout)
+                :else (flush on-send-error))))
+    [(fn [dimensions values]
+       (>!! report-ch [dimensions values]))
```

Comment:
This will block the reporter if we are sending stuff. which is probably not what we want.  We could add some kind of buffer.  Alternatively, we could also give it a transducer that merges incoming data with `collect`, so the channel would never have more than one queued 'message'.

Reply:
> Alternatively, we could give it a transducer that merges incoming data with collect, so the channel would never have more than one enqueued 'message' and a buffer of size 1 would be enough for it to never block senders.

How would this solve the problem? Specifically, the actual step of submitting can take a while, during which we could receive multiple attempts to write to the channel.

Reply:
I was confused about how transducers worked with channels.  I solved this issue (see my latest commit) by splitting the logic into two go blocks: one that collects and reduces the incoming reports, and one that pulls the reduced measurements and attempts to send them.  These two communicate via a channel.

With this setup we don't actually need channel buffers anywhere; RAM usage is bounded only by the `max-buffer-size` mechanism.